### PR TITLE
Always flush headers on first response write (#1202).

### PR DESCRIPTION
**Pull request by halter73**
#1202

### CODE REVIEW

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs [hunk 1]
```diff
@@ -782,30 +783,100 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
                         "");
-                    await connection.ReceiveEnd(
+                    await connection.Receive(
+                        "HTTP/1.1 200 OK",
+                        $"Date: {server.Context.DateHeaderValue}",
+                        "Content-Length: 12",
+                        "",
+                        "");
+
+                    flushed.Release();
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: Nice 👍

**Comment by cesarbs** [comment]
Check AppVeyor before merging.

**Comment by muratg** [comment]
@halter73 did you investigate the appveyor failure? otherwise is it good to check-in? I expect @pranavkm to try to get the build today.

File: test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs [hunk 2]
```diff
@@ -782,30 +783,100 @@ public Task ResponseStatusCodeSetBeforeHttpContextDisposedRequestMalformedReadIg
...
-        public async Task HeadResponseCanContainContentLengthHeaderButBodyNotWritten()
+        public async Task HeadResponseBodyNotWrittenWithAsyncWrite()
         {
+            var flushed = new SemaphoreSlim(0, 1);
+
             using (var server = new TestServer(async httpContext =>
             {
                 httpContext.Response.ContentLength = 12;
                 await httpContext.Response.WriteAsync("hello, world");
+                await flushed.WaitAsync();
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: Can you add reasonable timeouts to the semaphore Waits?

Comment (halter73) [inline test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/ResponseTests.cs]: These calls to WaitAsync() are in the web app on the test. If the web app hangs, the test will not hang. Instead the test will fail when it times out waiting for a response. The response timeout here is 1 minute like for most of our tests.

**Comment by halter73** [comment]
When the failing test (ThrowingSynchronousConnectionFilterDoesNotCrashServer) is run individually (using `dotnet test -method ...`) on the dev branch it fails in the same way [it did on appveyor](https://ci.appveyor.com/project/aspnetci/kestrelhttpserver/build/1.0.3750#L419). It seems to just be a flaky test. I'm modifying the test to make it more reliable.

**Comment by halter73** [comment]
@pranavkm Are you fine with me merging this?

**Comment by pranavkm** [comment]
:shipit:

**Pull request merged**